baseband/guppi/base.py

```diff
@@ -0,0 +1,360 @@
+
+        The routine uses the sample rate and number of samples per frame
+        from the first header in the file.
+
+        Returns
+        -------
+        frame_rate : `~astropy.units.Quantity`
+            Frames per second.
+        """
+        oldpos = self.tell()
+        self.seek(0)
+        try:
+            header = self.read_frame()
+            return (header.sample_rate /
+                    (header.samples_per_frame - header.overlap)).to(u.Hz)
+        finally:
+            self.seek(oldpos)
+
+
+class GUPPIFileWriter(VLBIFileBase):
+    """Simple writer/mapper for DADA files.
+
+    Adds `write_frame` and `memmap_frame` methods to the VLBI binary file
+    wrapper.  The latter allows one to encode data in pieces, writing to disk
+    as needed.
+    """
+    def write_frame(self, data, header=None, **kwargs):
+        """Write a single frame (header plus payload).
+
+        Parameters
+        ----------
+        data : `~numpy.ndarray` or `~baseband.guppi.GUPPIFrame`
+            If an array, a ``header`` should be given, which will be used to
+            get the information needed to encode the array, and to construct
+            the GUPPI frame.
+        header : `~baseband.guppi.GUPPIHeader`
+            Can instead give keyword arguments to construct a header.  Ignored
+            if ``data`` is a `~baseband.guppi.GUPPIFrame` instance.
+        **kwargs
+            If ``header`` is not given, these are used to initialize one.
+        """
+        if not isinstance(data, GUPPIFrame):
+            data = GUPPIFrame.fromdata(data, header, **kwargs)
+        return data.tofile(self.fh_raw)
+
+    def memmap_frame(self, header=None, **kwargs):
+        """Get frame by writing the header to disk and mapping its payload.
+
+        The header is written to disk immediately, but the payload is mapped,
+        so that it can be filled in pieces, by setting slices of the frame.
+
+        Parameters
+        ----------
+        header : `~baseband.guppi.GUPPIHeader`
+            Written to disk immediately.  Can instead give keyword arguments to
+            construct a header.
+        **kwargs
+            If ``header`` is not given, these are used to initialize one.
+
+        Returns
+        -------
+        frame: `~baseband.guppi.GUPPIFrame`
+            By assigning slices to data, the payload can be encoded piecewise.
+        """
+        if header is None:
+            header = GUPPIHeader.fromvalues(**kwargs)
+        header.tofile(self.fh_raw)
+        payload = GUPPIPayload.fromfile(self.fh_raw, memmap=True,
+                                        header=header)
+        return GUPPIFrame(header, payload)
+
+
+class GUPPIStreamBase(VLBIStreamBase):
+    """Base for GUPPI streams."""
+
+    _sample_shape_maker = GUPPIPayload._sample_shape_maker
+
+    def __init__(self, fh_raw, header0, squeeze=True, subset=()):
+
+        # GUPPI sample positions are determined by "packets", the bytesize of
```

Comment:
It isn't - for example `fh.samples_per_frame / fh._packets_per_frame` is 512 for a random sampling of Nikhil's data.  For his data, one complete sample is 16 bytes, but `PKTSIZE` is 8192.

Reply:
I wouldn't even bother with it other than `PKTIDX` is needed to double check offsets when new frames are read.

Reply:
This may be the only thing holding this up. I'd suggest to document it more clearly and for writing set a default of 8192 so that at least one doesn't have to put it in.

Reply:
Nikhil did more sleuthing and found that `PKTFMT` is the UDP data packet format (https://safe.nrao.edu/wiki/pub/Main/JoeBrandt/guppi_status_shmem.pdf), and that `FAST4K`, the other packet format besides `1SFA`, is only for total intensity.  I therefore agree we should set `PKTSIZE` to `8192`.  Less importantly, I also think we should default `PKTFMT` to `1SFA`, since that's linked to the packet size.